### PR TITLE
play, block, task: New attribute forks

### LINKED ISSUE #12170: support for "serial" on an individual task
(Feature Idea)

The natural way to handle configuration or other updates that require a rolling restart would be to perform updates in parallel, then notify a handler, which performs a restart with `serial`. But this is not possible, requiring either manual rolling restart or ugly hacks. See https://groups.google.com/forum/#!topic/ansible-project/rBcWzXjt-Xc

### LINKED ISSUE #24037: Add forks per play or per task
##### ISSUE TYPE

 - Feature Idea

##### COMPONENT NAME

core

##### ANSIBLE VERSION

v2.4

##### SUMMARY

We have the need to limit the number of tasks run in parallel for a specific task in order to not overload some infrastructure (in this case a terminal server). The total number of connections should not exceed 4.

Now we are using `serial:` for this, however this will do them 4 at a time, and it waits for the slowest one to be finished to do the next batch of 4. Which is suboptimal and slower, than using something like `forks: 4` where it fires off a new task as soon as one has finished.

For us this would make a lot of difference.

### PR DESCRIPTION
**Pull request by t-woerner**
With this patch it is possible to limit the number of concurrent task runs for the task where it has been added. This is similar to forks, but affects only one task. It can be used to serialize one or more tasks.

##### SUMMARY

The ability to limit concurrent task executions for a specific task would help [ansible-freeipa](https://github.com/freeipa/ansible-freeipa) to deploy a bigger number of replicas and dependant clients in a cluster environment in a nearly parallel way.

Right now it is not possible to install several replicas with [FreeIPA](https://freeipa.org/page/Main_Page) at the same time due to conflicts in the CA configuration step. For ansible-freeipa the installers have been split up into smaller pieces already to be able to replace some of the parts with new roles later on. With this patch in ansible it is possible to have a nearly parallel installation as only the affected step need to be executed in a serial way. The remaining tasks in the role can be done in parallel.

The creation of all replicas with ```forks:1``` is resulting in a by far longer deployment time. Also the proposals from [issue #12170](https://github.com/ansible/ansible/issues/12170) are not working as the registered output of previous tasks is needed.

This fixes #12170
This fixes #24037

##### ISSUE TYPE
 - Feature Pull Request

##### ANSIBLE VERSION

```
ansible 2.5.2
```

##### ADDITIONAL INFORMATION

Excerpt from the tasks/install.yml file used to install the replica with the new ```forks``` directive:

```yaml
...
  - name: Install - Setup CA
    ipareplica_setup_ca:
      ### server ###
      setup_ca: "{{ ipareplica_setup_ca }}"
      setup_kra: "{{ result_ipareplica_test.setup_kra }}"
      no_pkinit: "{{ ipareplica_no_pkinit }}"
      no_ui_redirect: "{{ ipareplica_no_ui_redirect }}"
      ### certificate system ###
      subject_base: "{{ result_ipareplica_prepare.subject_base }}"
      ### additional ###
      ccache: "{{ result_ipareplica_prepare.ccache }}"
      _ca_enabled: "{{ result_ipareplica_prepare._ca_enabled }}"
      _ca_file: "{{ result_ipareplica_prepare._ca_file }}"
      _ca_subject: "{{ result_ipareplica_prepare._ca_subject }}"
      _subject_base: "{{ result_ipareplica_prepare._subject_base }}"
      _pkinit_pkcs12_info: "{{ result_ipareplica_prepare._pkinit_pkcs12_info }}"
      _top_dir: "{{ result_ipareplica_prepare._top_dir }}"
      dirman_password: "{{ ipareplica_dirman_password }}"
      config_setup_ca: "{{ result_ipareplica_prepare.config_setup_ca }}"
      config_master_host_name: "{{ result_ipareplica_install_ca_certs.config_master_host_name }}"
      config_ca_host_name: "{{ result_ipareplica_install_ca_certs.config_ca_host_name }}"
      config_ips: "{{ result_ipareplica_prepare.config_ips }}"
    when: result_ipareplica_prepare._ca_enabled
    forks: 1
...
```

Parallel execution without `forks: 1`:
```
...
TASK [ipareplica : Install - Setup CA] *****************************************
task path: /root/ansible/ansible-freeipa/roles/ipareplica/tasks/install.yml:419
fatal: [ipareplica3.test3.local]: FAILED! => {"changed": false, "module_stderr": "Shared connection to ipareplica3.test3.local closed.\r\n", "module_stdout": "Traceback (most recent call last):\r\n  File \"/tmp/ansible_MJSMjy/ansible_module_ipareplica_setup_ca.py\", line 229, in <module>\r\n    main()\r\n  File \"/tmp/ansible_MJSMjy/ansible_module_ipareplica_setup_ca.py\", line 222, in main\r\n    ca.install(False, config, options, custodia=custodia)\r\n  File \"/usr/lib/python2.7/site-packages/ipaserver/install/ca.py\", line 223, in install\r\n    install_step_0(standalone, replica_config, options, custodia=custodia)\r\n  File \"/usr/lib/python2.7/site-packages/ipaserver/install/ca.py\", line 303, in install_step_0\r\n    use_ldaps=standalone)\r\n  File \"/usr/lib/python2.7/site-packages/ipaserver/install/cainstance.py\", line 467, in configure_instance\r\n    self.start_creation(runtime=runtime)\r\n  File \"/usr/lib/python2.7/site-packages/ipaserver/install/service.py\", line 520, in start_creation\r\n    run_step(full_msg, method)\r\n  File \"/usr/lib/python2.7/site-packages/ipaserver/install/service.py\", line 510, in run_step\r\n    method()\r\n  File \"/usr/lib/python2.7/site-packages/ipaserver/install/dogtaginstance.py\", line 473, in setup_admin\r\n    master_conn.simple_bind(self.admin_dn, self.admin_password)\r\n  File \"/usr/lib/python2.7/site-packages/ipapython/ipaldap.py\", line 1142, in simple_bind\r\n    bind_dn, bind_password, server_controls, client_controls)\r\n  File \"/usr/lib64/python2.7/contextlib.py\", line 35, in __exit__\r\n    self.gen.throw(type, value, traceback)\r\n  File \"/usr/lib/python2.7/site-packages/ipapython/ipaldap.py\", line 1030, in error_handler\r\n    raise errors.ACIError(info=\"%s %s\" % (info, desc))\r\nipalib.errors.ACIError: Insufficient access:  Invalid credentials\r\n", "msg": "MODULE FAILURE", "rc": 1}
fatal: [ipareplica4.test3.local]: FAILED! => {"changed": false, "module_stderr": "Shared connection to ipareplica4.test3.local closed.\r\n", "module_stdout": "Traceback (most recent call last):\r\n  File \"/tmp/ansible_muUnai/ansible_module_ipareplica_setup_ca.py\", line 229, in <module>\r\n    main()\r\n  File \"/tmp/ansible_muUnai/ansible_module_ipareplica_setup_ca.py\", line 222, in main\r\n    ca.install(False, config, options, custodia=custodia)\r\n  File \"/usr/lib/python2.7/site-packages/ipaserver/install/ca.py\", line 223, in install\r\n    install_step_0(standalone, replica_config, options, custodia=custodia)\r\n  File \"/usr/lib/python2.7/site-packages/ipaserver/install/ca.py\", line 303, in install_step_0\r\n    use_ldaps=standalone)\r\n  File \"/usr/lib/python2.7/site-packages/ipaserver/install/cainstance.py\", line 467, in configure_instance\r\n    self.start_creation(runtime=runtime)\r\n  File \"/usr/lib/python2.7/site-packages/ipaserver/install/service.py\", line 520, in start_creation\r\n    run_step(full_msg, method)\r\n  File \"/usr/lib/python2.7/site-packages/ipaserver/install/service.py\", line 510, in run_step\r\n    method()\r\n  File \"/usr/lib/python2.7/site-packages/ipaserver/install/dogtaginstance.py\", line 473, in setup_admin\r\n    master_conn.simple_bind(self.admin_dn, self.admin_password)\r\n  File \"/usr/lib/python2.7/site-packages/ipapython/ipaldap.py\", line 1142, in simple_bind\r\n    bind_dn, bind_password, server_controls, client_controls)\r\n  File \"/usr/lib64/python2.7/contextlib.py\", line 35, in __exit__\r\n    self.gen.throw(type, value, traceback)\r\n  File \"/usr/lib/python2.7/site-packages/ipapython/ipaldap.py\", line 1030, in error_handler\r\n    raise errors.ACIError(info=\"%s %s\" % (info, desc))\r\nipalib.errors.ACIError: Insufficient access:  Invalid credentials\r\n", "msg": "MODULE FAILURE", "rc": 1}
fatal: [ipareplica1.test3.local]: FAILED! => {"changed": false, "module_stderr": "Shared connection to ipareplica1.test3.local closed.\r\n", "module_stdout": "Traceback (most recent call last):\r\n  File \"/tmp/ansible_QegF3A/ansible_module_ipareplica_setup_ca.py\", line 229, in <module>\r\n    main()\r\n  File \"/tmp/ansible_QegF3A/ansible_module_ipareplica_setup_ca.py\", line 222, in main\r\n    ca.install(False, config, options, custodia=custodia)\r\n  File \"/usr/lib/python2.7/site-packages/ipaserver/install/ca.py\", line 223, in install\r\n    install_step_0(standalone, replica_config, options, custodia=custodia)\r\n  File \"/usr/lib/python2.7/site-packages/ipaserver/install/ca.py\", line 303, in install_step_0\r\n    use_ldaps=standalone)\r\n  File \"/usr/lib/python2.7/site-packages/ipaserver/install/cainstance.py\", line 467, in configure_instance\r\n    self.start_creation(runtime=runtime)\r\n  File \"/usr/lib/python2.7/site-packages/ipaserver/install/service.py\", line 520, in start_creation\r\n    run_step(full_msg, method)\r\n  File \"/usr/lib/python2.7/site-packages/ipaserver/install/service.py\", line 510, in run_step\r\n    method()\r\n  File \"/usr/lib/python2.7/site-packages/ipaserver/install/dogtaginstance.py\", line 473, in setup_admin\r\n    master_conn.simple_bind(self.admin_dn, self.admin_password)\r\n  File \"/usr/lib/python2.7/site-packages/ipapython/ipaldap.py\", line 1142, in simple_bind\r\n    bind_dn, bind_password, server_controls, client_controls)\r\n  File \"/usr/lib64/python2.7/contextlib.py\", line 35, in __exit__\r\n    self.gen.throw(type, value, traceback)\r\n  File \"/usr/lib/python2.7/site-packages/ipapython/ipaldap.py\", line 1030, in error_handler\r\n    raise errors.ACIError(info=\"%s %s\" % (info, desc))\r\nipalib.errors.ACIError: Insufficient access:  Invalid credentials\r\n", "msg": "MODULE FAILURE", "rc": 1}
changed: [ipareplica2.test3.local] => {"changed": true}
...
```
Serial execution with ```forks: 1```:
```
...
TASK [ipareplica : Install - Setup CA] *****************************************
task path: /root/ansible/ansible-freeipa/roles/ipareplica/tasks/install.yml:419
changed: [ipareplica1.test3.local] => {"changed": true}
changed: [ipareplica2.test3.local] => {"changed": true}
changed: [ipareplica3.test3.local] => {"changed": true}
changed: [ipareplica4.test3.local] => {"changed": true}
...
```

### CODE REVIEW

**Comment by dagwieers** [comment]
I would call this option **forks**, make it override the global **forks** option and make this configurable on a per-task, per-block, per-role and per-play option, see: https://github.com/ansible/ansible/issues/24037

**Comment by mattclay** [comment]
CI failure in Python 3.x unit tests: https://app.shippable.com/github/ansible/ansible/runs/73621/5/tests

The integration test failures are unrelated.

**Comment by notrix** [comment]
👍

**Comment by kustodian** [comment]
This looks like a really simple change for a feature that could be really useful. I have a few situations where this would make things a lot simpler.

If the core devs don't have anything against it, how can we get them to discuss about the name of the option and get it merged?

**Comment by bcoca** [comment]
@kustodian simple to add, but not when dealing with implications:

- how is this affected by global fork limits?
- how does this interact with 'serial' ?
- how does this affect max_fail_percentage?
- what is the expected result with run_once or 'bypass host loop' actions?
- how is delegation affected?
- at which failure point do we trigger block error handling (rescue)? 
- how do we handle variable/value precedence when they return facts?
- how do we handle precedence when they update inventory or other shared resources (add_host/group_by)?

And I'm sure I'm missing quite a few considerations, the list above is just from the top of my head. These, and more, have been brought up in adhoc and core IRC meetings discussing this topic (but possibly not this specific ticket) and we have not found good answers for most of those, we need to be able to answer all of the above before we consider adding it to the code base.

**Comment by kustodian** [comment]
@bcoca then it's best to just close this ticket with the explanation that it complicates a lot of things and because of that it will not be added into Ansible anytime in the near future.

**Comment by bcoca** [comment]
@kustodian we are still considering the feature, we just need to figure out how to deal with all the associated issues it entails.

**Comment by t-woerner** [comment]
@bcoca here are answers for the items where I am able to provide answers for the open questions.

> @kustodian simple to add, but not when dealing with implications:
> 
>     * how is this affected by global fork limits?
> 
The minimum of global fork limit and inventory hosts defines how many workers are created. The current implementation of max_concurrent is done in StrategyBase._queue_task. It is only effective if max_concurrent is bigger than 0 and max_concurrent is smaller than the number of workers. If the current worker index (self._cur_worker) is greater or equal task.max_concurrent the index is set back to 0. This is the same mechanism that is used to switch back to worker 0 if the maximum number of workers has been reached.
That means that max_concurrent is not able to increase the number of workers. A decrease of the global fork limit will also decrease the maximum number that can be used with max_concurrent to have an effect. max_concurrent could be seen as a task local limitation of the number of workers.

>     * how does this interact with 'serial' ?
> 
Batches are created according to the number of hosts that need to be processed with limiting the number of hosts in a batch to serial count. These batches are processed one after the other. The inventory is restricted to the hosts that are part of the batch while processing the batch. As the number of hosts defined in the inventory will also limit the number of workers, max_concurrent will only have an effect if it is smaller than serial if serial is used. serial is limiting the number of workers, that could be further limited with max_concurrent.

>     * how does this affect max_fail_percentage?
> 
The calculation of the percentage of fails is done according to the iterator.batch_size and the number of failed hosts. As iterator.batch_size and also failed_hosts is not changed by the implementation of max_concurrrent I do not see how this could be affected.

>     * what is the expected result with run_once or 'bypass host loop' actions?
> 
As max_concurrent is only temporary limiting the number of workers without changing the task or the batch size I do not see how this could affect run_once as there is no way to use max_concurrent to force to not use any workers. The expected result is the same as without max_concurrent.

>     * how is delegation affected?
> 
I do not see how max_concurrent with the temporary limitation of workers should affect this. Is global fork limit affecting it?

>     * at which failure point do we trigger block error handling (rescue)?
> 
I do not see how max_concurrent with the temporary limitation of workers should affect this. Is global fork limit affecting it?

>     * how do we handle variable/value precedence when they return facts?
> 
I do not see how max_concurrent with the temporary limitation of workers should affect this. Is global fork limit affecting it?

>     * how do we handle precedence when they update inventory or other shared resources (add_host/group_by)?
> 
I do not see how max_concurrent with the temporary limitation of workers should affect this. Is global fork limit affecting it?

> 
> And I'm sure I'm missing quite a few considerations, the list above is just from the top of my head. These, and more, have been brought up in adhoc and core IRC meetings discussing this topic (but possibly not this specific ticket) and we have not found good answers for most of those, we need to be able to answer all of the above before we consider adding it to the code base.

Please let us start an open discussion.

**Comment by dagwieers** [comment]
/me still thinks this ought to be a per-playbook, per-play, per-block, per-task directive named **forks**.
Which would influence the global/command-line forks option.

**Comment by t-woerner** [comment]
> @t-woerner that is mostly how this implementation works, but the discussion is more about 'how should it work?'
> 
> I do think yours is one of the simplest approaches by having it be a subset of existing workers, but I expect users to do something like the following: serial: 3 forks: 5 max_concurrent: 5 and expect an expansion on serial. Even with current code, they should at least get a warning that you can never exceed min(serial, forks) .. though i'm not as sure as you are about the actual behaviour (need to test).
> 
I used the name max_concurrent exactly because of the current behaviour. The number of workers is limited by forks and also the number of hosts in the current batch. I would not expect a simply expansion in this case. The serial option might have been added because of an issue with controller, nodes, used software or environment. This is the same as max_concurrent. I added this to the role to be able to successfully deploy replicas. Right now it is 2 using the same master without getting into trouble, but with some enhancements this might increase to higher values. This can then be adapted with the version of the software for example. But it should not automatically be increased.

Information about a low serial setting with a high forks setting might be good to have. This might also be good for settings in other levels. But most likely people will only see this while debugging.

> Another example, when combining max_concurrent: X with run_once ... i would also expect a warning or error since those settings are in conflict and we can only obey one. Its simpler with serial/fork since run_once is clearly more specific (on the task vs play and config) but when both are on the task there is no clear way to set expectations (see loop/when issues).
> 
I think it should be possible to add a conflict test with run_once on the task level, but it will not be simple to add a test like this by looking at other levels.

> It is not only mapping out 'desired behaviour' its also responding to conflicts and giving the user feedback about incompatibilities.
> 
Testing and giving feedback of all settings of all levels (command line, play, blocks, tasks, ...) that might conflict or influence others is really way more work and also invasive than this PR is supposed to be and to be honest it is out of scope of this PR in my opinion. It is something global that needs to be started independently.

> FYI, the discussion is always open, it just has lacked someone pushing it to the forefront.

**Comment by maxamillion** [comment]
> /me still thinks this ought to be a per-playbook, perl-role, per-play, per-block, per-task directive named **forks**. Which would influence the global/command-line forks option. Could even be **ansible_forks** if need be.

I'm in favor of this, I don't care about the exact name so long as it is or includes `forks` as I think it would help alleviate some of the confusion of "what's this do?" because of consistent nomenclature with the global config.

Aside from bikeshedding the name, FWIW, I'm a fan of the feature and the implementation. It's as simple as I could have hoped such a change would be, it's easy to track what's going on here. Needs docs, tests, and a changelog fragment but otherwise I'm 👍

**Comment by dagwieers** [comment]
> Aside from bikeshedding the name, FWIW, I'm a fan of the feature and the implementation.

It currently only implements it on the task-level though, and it seems to behave like `serial:` and not like `forks:`. So if the batch has not completed because one system is stalling, everything is stalling, the remaining 4 slots are not being used...

**Comment by t-woerner** [comment]
I have updated the PR, the attribute has been renamed to forks and it can be used in all levels of the chain: play, blocks, and tasks.

**Comment by maxamillion** [comment]
LGTM 👍

**Review by dagwieers — APPROVED** [review]
LGTM, I have tested this extensively and it appears to behave exactly as I expected.

**Comment by alikins** [comment]
Needs tests.

**Comment by dagwieers** [comment]
Ansible is configured by default to have 5 forks. People may have modified the number of forks (or supplied a different number using `-f` on the commandline), and this value takes priority even if you use a higher setting of `forks:` on a play, block, task level it will use the configured (provided on CLI) forks-number.

Personally I think that is fine, but it may be confusing to users, so it could be useful to expose a warning if the user has a lower number of forks configured. Not sure if we want this, or not.

**Comment by dagwieers** [comment]
@alikins I was wondering about that as well, but I don't know how to test if the behaviour is correct. My tests included 20 localhost entries with a default sleep of 1, and some have a sleep of 120 seconds. I wanted to see if it would continue using all (remaining) fork-slots if some of them take longer to complete (this is how `serial:` works). I don't know how you would test this.

**Comment by t-woerner** [comment]
Yes, the test is a question. I have also created a test yml file with sleeps and 12 localhost entries. I can add it, but I am not sure that it will be useful because you need to have a look to see that the processing is correct according to the forks defined in different layers of the chains.
It would only be possible to verify the warning with run_once usage and also the forks setting that will written out to the debug log. But all that in different files to make sure that the proper test will be verified.

**Comment by dagwieers** [comment]
This PR also helps to avoid any tasks that fails to work concurrently (like some APIs, as Cobbler, or modifying the same file on localhost). Another use-case is related to a terminal server that has a limited number of slots that can be addressed at any given time. (Using serial slows down and has other side-effects)

**Review by gundalow — CHANGES_REQUESTED** [review]
If this is accepted we will need docs.
Could test it via `runme.sh` to run a Playbook and count the number of forks found in the debug logs

**Comment by kustodian** [comment]
Any updates about this feature?

**Comment by dagwieers** [comment]
@t-woerner We are adding file locking to modules to support concurrent access, and had a similar issue. You can find our tests at: https://github.com/ansible/ansible/pull/52567

And maybe you can do something similar as we do there to count the number of concurrent instances by locking and releasing the lock. But we only have a few weeks left to get this in Ansible v2.8...

File: lib/ansible/plugins/strategy/__init__.py
Comment (bcoca) [inline lib/ansible/plugins/strategy/__init__.py]: i dont think this requires a warning, since min forks is 1 and run_once will be the 'min'

File: lib/ansible/plugins/strategy/__init__.py
Comment (bcoca) [inline lib/ansible/plugins/strategy/__init__.py]: we might want to warn 'forks is ignored due to strategy == free'

Comment (bcoca) [reply lib/ansible/plugins/strategy/__init__.py]: hmm, its implemented in free strat, unsure why you need to check for free here then

Comment (t-woerner) [reply lib/ansible/plugins/strategy/__init__.py]: The code in StrategyBase limits the amount of workers, which is the proper solution for linear. For the free strategy the workers should not get limited as all the other workers might be used for other tasks.

Comment (t-woerner) [reply lib/ansible/plugins/strategy/__init__.py]: Please have a look at the third patch for forks support with the free strategy.

File: lib/ansible/plugins/strategy/__init__.py
Comment (t-woerner) [reply lib/ansible/plugins/strategy/__init__.py]: I can remove this again, but you have been asking for this before: "Another example, when combining max_concurrent: X with run_once ... i would also expect a warning or error since those settings are in conflict and we can only obey one. Its simpler with serial/fork since run_once is clearly more specific"

Comment (bcoca) [reply lib/ansible/plugins/strategy/__init__.py]: same as run_once and serial and --forks, it takes precedence, as long as that is clearly documented  about the keywords i dont see need for runtime warning

Comment (dagwieers) [reply lib/ansible/plugins/strategy/__init__.py]: Agreed.

Comment (t-woerner) [reply lib/ansible/plugins/strategy/__init__.py]: O.k, these messages are now a debug messages.

**Comment by dagwieers** [comment]
Let's not delay getting this in.

**Review by dagwieers — APPROVED** [review]
LGTM

**Comment by t-woerner** [comment]
There is now a test case that is used for linear and free strategy.

**Review by bcoca — APPROVED** [review]

**Comment by nitzmahone** [comment]
Thanks for continuing to move this forward.... With the fork/threading + process-model plugin stuff that's coming in 2.9 (it's currently scheduled to merge right after we branch stable-2.8 on March 21), I'm thinking this might need a revisit right after that's merged.

I'm +1 for the feature itself, though I actually preferred the original name of `max_concurrent`, as it's process-model agnostic (`forks` is a misnomer in a threaded world, and something we'll have to decide how to address more generally across the project as threading becomes mainstream).

The tests also aren't fully validating the interactions between this and `serial`, `run_once`, etc.; especially with the stuff that's coming early in 2.9, we'll want to ensure that those interactions are both well-defined and well-tested, since there will be many more places for it to potentially break.

Finally, docs will be very important for this feature, and should be included in the PR once we're all agreed about the naming and timing.

**Comment by bcoca** [comment]
fork applies to BOTH processes and threads

**Comment by sivel** [comment]
I'm going to have to agree with @nitzmahone, we should be careful to create further confusion using the name `forks`.  In the python world, which is what we deal with for developing Ansible, forks are only synonymous with multiprocessing, and the new process model plugins will break that relationship.

We are likely going to push to deprecating the use of the word `forks`, in favor or other terminology more suited for their use cases.

Additionally, the use of `forks` here is misleading, as it doesn't actually affect the number of running forks, just limits the number of those running forks that are utilized.  I think `max_concurrent` for this use makes a lot of sense, and the `--forks` CLI arg, would be better renamed as `--max-workers`.

**Review by sivel — CHANGES_REQUESTED** [review]
Just putting in my general "request changes" until this can be reconciled with the future of process models, and decoupling terminology away from the underlying process model.

**Comment by dagwieers** [comment]
Please use the same name for the option and the playbook directive. So if you go for `--max-workers` then call the directive `max_workers`. For clarity. That's why I preferred `forks` over `max_concurrent`.

If it is doing the same thing, name it the same or you add to the confusion.

**Comment by sivel** [comment]
As I described they don't do the same thing. As of now `--forks` dictates the max number of workers that will be spawned, whereas `task.fork` limits the count of those workers that will accept work. Inherently they are different things.

File: lib/ansible/playbook/base.py
Comment (bcoca) [inline lib/ansible/playbook/base.py]: needs to be ` static=True` as templating per host makes no sense here.

**Review by bcoca — CHANGES_REQUESTED** [review]

**Comment by dagwieers** [comment]
@sivel Sure internally, but the effect to the user is the number of workers being used. Good interface design does not expose irrelevant internals to the user IMO.

File: lib/ansible/playbook/base.py
Comment (t-woerner) [reply lib/ansible/playbook/base.py]: I have marked the forks attribute as static. Interesting is that there is only one attribute so far that is defined to be static: The register attribute.

**Comment by t-woerner** [comment]
What priority do you see for this PR? If it is going to be part of 2.8, then it has high priority and I will be able to spend more time. But if it needs to wait till 2.9 or later, then it is low priority and will move down to the bottom of my work plan.

**Comment by dagwieers** [comment]
@t-woerner I hope this can ship with Ansible v2.8.

**Comment by sivel** [comment]
To be completely honest, this may not ship with 2.8.  We are evaluating the further impact of this with a new feature for selecting the process model (threading vs forking) which is in progress and intended to land in 2.9.  The impact of this work could actually cause issues with that planned feature, which could be negative if accepted first.

File: test/integration/targets/forks/test_forks.yml
Comment (sivel) [inline test/integration/targets/forks/test_forks.yml]: Instead of `python` this should be utilizing `ansible_python_interpreter`, potentially falling back to another python.

Comment (bcoca) [reply test/integration/targets/forks/test_forks.yml]: ansible_playbook_python, since you are executing on "localhosts" (i assume connection = local)

**Comment by kustodian** [comment]
Any updates about threading vs forking?

**Comment by sivel** [comment]
> Any updates about threading vs forking?

To my knowledge, we are working through some testing decisions about how we will properly test both process models without overloading the CI infrastructure, but still get enough coverage.

File: test/integration/targets/forks/test_forks.yml
Comment (t-woerner) [reply test/integration/targets/forks/test_forks.yml]: I have fixed this in the test case.

**Comment by t-woerner** [comment]
According to https://docs.ansible.com/ansible/latest/roadmap/ROADMAP_2_9.html the Beta 1 Feature (No new functionality (including modules/plugins) to any code) is in less than a month.

Will the threading and forking work get merged for 2.9 still? There has not been an update for https://github.com/ansible/ansible/pull/44280 since June 21st.

**Review by bcoca — APPROVED** [review]

File: test/integration/targets/forks/test_forks.yml
Comment (sivel) [inline test/integration/targets/forks/test_forks.yml:20]: Per @bcoca's comment, this should be `ansible_playbook_python` instead of `ansible_python_interpreter` since we are just running against localhost

**Review by sivel — DISMISSED** [review]

File: test/integration/targets/forks/test_forks.yml
Comment (t-woerner) [reply test/integration/targets/forks/test_forks.yml:20]: The inventory contains this:
`ansible_python_interpreter="{{ ansible_playbook_python }}"`

It is done this way in lots of inventory files in the test folder.

**Comment by t-woerner** [comment]
@gundalow 

> If this is accepted we will need docs.

When is this needed and what is the format?

> Could test it via `runme.sh` to run a Playbook and count the number of forks found in the debug logs

There are ests for linear and free strategy: https://github.com/ansible/ansible/pull/42528/commits/e31cc8d7a67b41351fbeb561f0af0edc5479d94f

**Comment by jimi-c** [comment]
Merged https://github.com/ansible/ansible/pull/60702 which supersedes this.

**Pull request closed without merging**
